File: solutions/Easy/Symmetric_Tree_0101.py
from typing import Optional
from collections import deque

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        """
        Function: isSymmetric
        Description: Checks if a binary tree is symmetric around its center.

        Parameters:
        - root (Optional[TreeNode]): The root of the binary tree.

        Returns:
        - bool: True if the tree is symmetric, otherwise False.
        """
        if not root:
            return True

        def isMirror(t1: Optional[TreeNode], t2: Optional[TreeNode]) -> bool:
            """
            Helper function to check if two subtrees are mirror images of each other.
            """
            if not t1 and not t2:
                return True
            if not t1 or not t2 or t1.val != t2.val:
                return False
            return isMirror(t1.left, t2.right) and isMirror(t1.right, t2.left)

        return isMirror(root.left, root.right)

    def isSymmetricIterative(self, root: Optional[TreeNode]) -> bool:
        """
        Iterative approach using a queue.
        """
        if not root:
            return True

        queue = deque([(root.left, root.right)])

        while queue:
            t1, t2 = queue.popleft()

            if not t1 and not t2:
                continue
            if not t1 or not t2 or t1.val != t2.val:
                return False

            queue.append((t1.left, t2.right))
            queue.append((t1.right, t2.left))

        return True


# Helper functions for testing
def list_to_binary_tree(lst):
    """Converts a level-order list to a binary tree."""
    if not lst or lst[0] is None:
        return None

    nodes = [TreeNode(val) if val is not None else None for val in lst]

    for i, node in enumerate(nodes):
        if node:
            left_idx = 2 * i + 1
            right_idx = 2 * i + 2
            if left_idx < len(lst):
                node.left = nodes[left_idx]
            if right_idx < len(lst):
                node.right = nodes[right_idx]

    return nodes[0]  # Root node


# Example Test Cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1: Symmetric tree
    root = list_to_binary_tree([1, 2, 2, 3, 4, 4, 3])
    print(solution.isSymmetric(root))  # Expected output: True
    print(solution.isSymmetricIterative(root))  # Expected output: True

    # Test case 2: Asymmetric tree
    root = list_to_binary_tree([1, 2, 2, None, 3, None, 3])
    print(solution.isSymmetric(root))  # Expected output: False
    print(solution.isSymmetricIterative(root))  # Expected output: False

    # Test case 3: Single node tree
    root = list_to_binary_tree([1])
    print(solution.isSymmetric(root))  # Expected output: True
    print(solution.isSymmetricIterative(root))  # Expected output: True

    # Test case 4: Empty tree
    root = list_to_binary_tree([])
    print(solution.isSymmetric(root))  # Expected output: True
    print(solution.isSymmetricIterative(root))  # Expected output: True

    # Test case 5: Larger symmetric tree
    root = list_to_binary_tree([1, 2, 2, 3, 4, 4, 3, None, None, 5, 6, 6, 5])
    print(solution.isSymmetric(root))  # Expected output: True
    print(solution.isSymmetricIterative(root))  # Expected output: True

    # Test case 6: Larger asymmetric tree
    root = list_to_binary_tree(
        [1, 2, 2, 3, 4, 4, 3, None, None, 5, 6, 6, None, 5])
    print(solution.isSymmetric(root))  # Expected output: False
    print(solution.isSymmetricIterative(root))  # Expected output: False

    print("All test cases passed!")

# Complexity Analysis
# Time Complexity: O(N), where N is the number of nodes in the binary tree. We traverse the entire tree once, so the time complexity is O(N).

# Space Complexity: O(N) for the recursive approach and O(N) for the iterative approach.
# In the recursive approach, the maximum depth of the recursion stack is the height of the tree, which is O(N) in the worst case.
# In the iterative approach, the queue can contain up to O(N) nodes in the worst case.

File: solutions/Easy/Symmetric_Tree_0101.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val);
 *     this.left = (left===undefined ? null : left);
 *     this.right = (right===undefined ? null : right);
 * }
 */

/**
 * Function: isSymmetric (Recursive Approach)
 * Description: Given the root of a binary tree, this function checks if the tree is 
 * symmetric (a mirror image of itself) using recursion.
 *
 * @param {TreeNode} root - The root node of the binary tree.
 * @returns {boolean} - True if the tree is symmetric, otherwise false.
 */
const isSymmetric = function(root) {
    if (!root) return true;

    /**
     * Helper function to check if two subtrees are mirror images.
     * @param {TreeNode} left - Left subtree.
     * @param {TreeNode} right - Right subtree.
     * @returns {boolean} - True if they are mirror images, otherwise false.
     */
    const isMirror = (left, right) => {
        if (!left && !right) return true; // Both nodes are null
        if (!left || !right) return false; // One is null, but not the other
        if (left.val !== right.val) return false; // Values don't match

        // Check symmetry in subtrees: left's left with right's right and left's right with right's left
        return isMirror(left.left, right.right) && isMirror(left.right, right.left);
    };

    return isMirror(root.left, root.right);
};

/**
 * Function: isSymmetricIterative (Iterative Approach)
 * Description: Uses BFS (queue) to check for symmetry in a binary tree.
 *
 * @param {TreeNode} root - The root node of the binary tree.
 * @returns {boolean} - True if the tree is symmetric, otherwise false.
 */
const isSymmetricIterative = function(root) {
    if (!root) return true;

    let queue = [[root.left, root.right]];

    while (queue.length > 0) {
        let [left, right] = queue.shift();

        if (!left && !right) continue; // Both null, continue checking
        if (!left || !right || left.val !== right.val) return false; // One null or values mismatch

        // Enqueue pairs to check symmetry
        queue.push([left.left, right.right]); // Outer symmetry
        queue.push([left.right, right.left]); // Inner symmetry
    }

    return true;
};

// Example Test Cases
if (require.main === module) {
    const TreeNode = function(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    };

    console.log("Test 1: Symmetric tree");
    let root1 = new TreeNode(1, 
        new TreeNode(2, new TreeNode(3), new TreeNode(4)), 
        new TreeNode(2, new TreeNode(4), new TreeNode(3))
    );
    console.log(isSymmetric(root1)); // Expected output: true
    console.log(isSymmetricIterative(root1)); // Expected output: true

    console.log("Test 2: Asymmetric tree");
    let root2 = new TreeNode(1, 
        new TreeNode(2, null, new TreeNode(3)), 
        new TreeNode(2, null, new TreeNode(3))
    );
    console.log(isSymmetric(root2)); // Expected output: false
    console.log(isSymmetricIterative(root2)); // Expected output: false

    console.log("Test 3: Single node tree");
    let root3 = new TreeNode(1);
    console.log(isSymmetric(root3)); // Expected output: true
    console.log(isSymmetricIterative(root3)); // Expected output: true

    console.log("Test 4: Left-skewed tree");
    let root4 = new TreeNode(1, new TreeNode(2, new TreeNode(3)));
    console.log(isSymmetric(root4)); // Expected output: false
    console.log(isSymmetricIterative(root4)); // Expected output: false

    console.log("Test 5: Right-skewed tree");
    let root5 = new TreeNode(1, null, new TreeNode(2, null, new TreeNode(3)));
    console.log(isSymmetric(root5)); // Expected output: false
    console.log(isSymmetricIterative(root5)); // Expected output: false

    console.log("Test 6: Empty tree");
    console.log(isSymmetric(null)); // Expected output: true
    console.log(isSymmetricIterative(null)); // Expected output: true
}

/*
Time Complexity:
- Recursive: O(n) - Each node is visited once.
- Iterative: O(n) - Each node is processed once.

Space Complexity:
- Recursive: O(h) - Recursion depth (O(log n) for balanced, O(n) for skewed trees).
- Iterative: O(n) - Queue stores nodes level-wise.
*/

export { isSymmetric, isSymmetricIterative };
